### DISCUSSION TITLE
[schema question] artificially combining data model to reduce multiple http hits

### DISCUSSION THREAD

**Discussion by guykatz** [graphql-java/graphql-java#2233] [Q&A]
Hi all;
after designing my schema I have a few disparate subsystems without refference from one to another.
I have noticed that in my UI application, I need to perform multiple http hits in order to get data. so essentially ,I solved the over/under fetching but still perform many http hits (like in REST).
is there a best practice for this? should I "artificially" add a root node connecting my disparate systems in order to allow a query that will be able to return all the data at once?
any comments/suggestions would be greatly appreciated.
thanks!

**Comment by ooga** [comment]
Not really sure to have well understood but GraphQL federation might be a solution. 
Some documentation : https://www.apollographql.com/docs/federation/managed-federation/overview/

Some Java & Kotlin implementations can be found at https://www.apollographql.com/docs/federation/other-servers/

**Comment by bbakerman** [comment]
I cant quite envisage your problem.  Ban you outline a mini schema (in psudeo form) to show your problem please ?

**Reply by guykatz** [reply]
thanks guys here's a very simplified example I am lookoing for a best practice for. considering the following schema:
```
type user{
    name:string!
}
type product{
 price: Int!
}
queries{
  users: [user!]!
  products: [product!]!
}
```
so users and products have 2 different queries. if I need to display info in a page from both I will need to do 2 queries (usually 2 http hits or should I bulk?). as the schema grows and I need to display more data from more sub part of the schema I will need to do more HTTP hits.
when I say artificially connect the two I mean the following:
```
type system{
  users: [User!]
  products: [Products!]
}
query{
system: System! 
}
```
now I can ask for users and products in the same query if I need.
so essentially I am looking to understand whats the best practice in these kind of situations
thanks in advance

**Reply by lqc** [reply]
You don't need the fake "System" type. All queries already have a root type "Query":

```
query {
  users { name }
  products { price }
}
```
Will work just fine with the first schema.

**Reply by bbakerman** [reply]
Yeah as Lukaz says above it will work find.  One of the things to understand about graphql is that it is a "field execution" mechanism.

A query with 2 fields will invoke the DataFecther on each field -  So one query can get N lots of entity data. (unlike REST say)

However I am wondering if you meant "combine" the 2 calls (1 to get user data and 1 to get system data) into one logical call in the backend?

if you meant this (and I say if because i am not sure) then there is a way to know if 2 or more fields have been asked for.

This is called field subselection - see https://www.graphql-java.com/documentation/v16/fieldselection/

This of course supposes that you have a HTTP api you can call that gives out both users and system in one call.

**Reply by guykatz** [reply]
thanks guys for all the answer. I think I know what to do. probably @lqc 's answer fits me best - I overlooked this option.
thanks @bbakerman for the ellaborate answer

**Comment by nutharsh** [comment]
` I have a few disparate subsystems without refference from one to another. `

Can you elaborate what do you mean by this?
Do you mean separate services or different entities within the same deployable service?
